### ISSUE TITLE
Mock S3 tests instead of using s3-server

### ISSUE THREAD

**Issue by brainstorm** [umccr/htsget-rs#131]
See issues: 

* https://github.com/datenlord/s3-server/issues/108
* https://github.com/s3d-rs/s3d/issues/16

This only affects testing, but it pulls in vulnerable Rust deps and also OpenSSL which we really don't need nor want.

**Comment by mmalenic** [comment]
Yeah, I think this one is important, as we would like to use rustls for everything if possible.

**Comment by mmalenic** [comment]
Currently the only place this comes up is here: 
https://github.com/umccr/htsget-rs/blob/4ed6d8ed81f219040b553b77091dd73527e43e6b/htsget-search/src/storage/aws.rs#L247-L252

If we wanted to remove this without any additional dependencies, we could refactor calls to the aws sdk into a trait, and a mock test implementation to only test the logic of our code, as suggested [here](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/testing.html).

**Comment by brainstorm** [comment]
Indeed, we should just mock instead of running a server in the background, although isn't mocking per-method a bit tedious?

This issue outlines how to mock at a ::Client level, perhaps easier:

https://github.com/awslabs/aws-sdk-rust/issues/199

**Comment by Nugine** [comment]
`s3-server` has ended its maintenance as I announced recently in the tracking issue. https://github.com/datenlord/s3-server/issues/108#issuecomment-1384907255

[`s3s`](https://github.com/Nugine/s3s) has fixed the vulnerable deps problem because it uses custom codegen instead of depending on the unmaintained Rusoto.
[Here](https://github.com/Nugine/s3s/blob/1b0d935cc7859f4d9b37431adff4f63c42e0b809/crates/s3s-fs/src/main.rs) is the example if you have interest in migrating to [`s3s-fs`](https://github.com/Nugine/s3s). It would be easy as the interface is almost the same.

I'm also thinking how to provide a simple method to mock a full S3 client. The auto-generated [S3 trait](https://github.com/Nugine/s3s/blob/1b0d935cc7859f4d9b37431adff4f63c42e0b809/crates/s3s/src/ops/generated.rs#L13-L502) may give you a kind of inspiration. [Here](https://github.com/Nugine/s3s/blob/1b0d935cc7859f4d9b37431adff4f63c42e0b809/crates/s3s-fs/src/s3.rs) is an example of how to implement the trait.

**Comment by brainstorm** [comment]
Thanks for the detailed response, @Nugine! I'd be very interested in full S3 client mocking since I'd like to avoid having s3 servers running when running the testsuite (less complexity and less problems with deployment, operations, ports, etc..).

**Comment by brainstorm** [comment]
@Nugine, I'm wondering if https://crates.io/crates/s3s-aws is your ongoing effort to "mock a full S3 client", I'm a bit on a holding pattern to see what to use that doesn't require a running s3 server in the background :)

**Comment by Nugine** [comment]
[s3s-aws](https://docs.rs/s3s-aws/0.2.0/s3s_aws/index.html) provides two different types: [`Connector`](https://docs.rs/s3s-aws/0.2.0/s3s_aws/struct.Connector.html) and [`Proxy`](https://docs.rs/s3s-aws/0.2.0/s3s_aws/struct.Proxy.html).

`s3s_aws::Connector` implments [`SmithyConnector`](https://docs.rs/aws-smithy-client/0.54.1/aws_smithy_client/bounds/trait.SmithyConnector.html). It can be used to turn an [`S3Service`](https://docs.rs/s3s/0.2.0/s3s/service/struct.S3Service.html) into [the http connector](https://docs.rs/aws-types/0.54.1/aws_types/sdk_config/struct.Builder.html#method.http_connector) in [`SdkConfig`](https://docs.rs/aws-types/0.54.1/aws_types/sdk_config/struct.SdkConfig.html). So you just need to change the sdk config in mocking tests.

[Here](https://github.com/Nugine/s3s/blob/bf1f0143ce8d9d93a17d3aa804df0b2da3d6c639/crates/s3s-fs/tests/it_aws.rs) is an example combining `aws-sdk-s3`, `s3s-aws` and `s3s-fs`.

The logical call stack is like this:

```
aws_sdk_s3::Client::get_object
    s3s_aws::Connector::call
            s3s::S3Service::call
                <s3s_fs::FileSystem as s3s::S3>::get_object
```

It does not bind any port. The whole action happens in the test process. 

Note that [`s3s_fs::FileSystem`](https://docs.rs/s3s-fs/0.2.0/s3s_fs/struct.FileSystem.html) changes the data directory. You may need to serialize the tests to avoid race cases.

However, `s3s_fs::FileSystem` is not a full-featured and [well-tested](https://github.com/Nugine/s3s/issues/4) implementation. It may be buggy. If you want to avoid the possible bugs inside `s3s_fs::FileSystem`, I would recommand testing against [minio](https://github.com/minio/minio), an S3-compatible object storage server.

**Comment by brainstorm** [comment]
Thanks a ton @Nugine, I just migrated the tests and they seem to work... could you please publish your latest 0.2.1-dev changes from your master so that I can pin it to a version from crates.io? I'm pointing it to your git repo for the moment on PR #150 but it'll probably break soon if it's not pinned.

**Comment by Nugine** [comment]
I released v0.3.0 just now. It contains several bug fixes and a small break change. Feel free to open issues if you find any problem.
